### PR TITLE
New Redis Client API

### LINKED ISSUE #21108: How to use Redis pub/sub with Quarkus and ReactiveRedisClient?
I am trying to use Redis pub/sub with Quarkus, but I cannot find any documentation about it. I managed to use the Quarkus Redis Extension with Redis Streams xadd and xread, but cannot make Redis pub/sub work. I am using **ReactiveRedisClient** and when I try to subscribe to a redis channel in a Quarkus test I receive following error:

`Error received: io.vertx.core.impl.NoStackTraceThrowable: PubSub command in connection-less mode not allowed`

I found the error is thrown in vertx redis extension **BaseRedisClient** class:
```
@Override
  public Future<@Nullable Response> send(Request command) {
    final Promise<Response> promise = vertx.promise();

    if (command.command().isPubSub()) {
      // mixing pubSub cannot be used on a one-shot operation
      promise.fail("PubSub command in connection-less mode not allowed");
      return promise.future();
    }
  }
```

Generally I find Quarkus would benefit of documenting how to use Redis Pub/Sub cause I found other people are also confused  how to use it exactly.

Thanks in advance for any help!

### LINKED ISSUE #24056: ReactiveRedisClient: Transaction with multi() not working
### Describe the bug

I'm trying to implement the check-and-set pattern (https://redis.io/topics/transactions#cas) using the ReactiveRedisClient. But commands after the `multi` command do not seem to get `queued`, they're executed directly. When using the blocking version, everything works as expected.

When using the following code, the log output of the reactive endpoint is:
MULTI OK
SET OK
EXEC []

The output of the blocking endpoint is:
MULTI OK
SET QUEUED
EXEC [OK]

Code:

```
import io.quarkus.redis.client.RedisClient;
import io.quarkus.redis.client.reactive.ReactiveRedisClient;
import io.smallrye.mutiny.Uni;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.inject.Inject;
import javax.ws.rs.GET;
import javax.ws.rs.Path;
import java.util.List;

@Path("/")
public class ReactiveGreetingResource {
  private final Logger log = LoggerFactory.getLogger(ReactiveGreetingResource.class);

  @Inject
  RedisClient redisClient;

  @Inject
  ReactiveRedisClient reactiveRedisClient;

  @GET
  @Path("/reactive")
  public Uni<String> reactive() {
    return this.reactiveRedisClient
      .multi()
      .flatMap(response -> {
        log.info("MULTI {}", response);
        return this.reactiveRedisClient.set(List.of("a", "b"));
      })
      .flatMap(response -> {
        log.info("SET {}", response);
        return this.reactiveRedisClient.exec();
      })
      .map(response -> {
        log.info("EXEC {}", response);
        return "done";
      });
  }

  @GET
  @Path("/blocking")
  public String blocking() {
    var multi = redisClient.multi();
    log.info("MULTI {}", multi);

    var set = redisClient.set(List.of("a", "b"));
    log.info("SET {}", set);

    var exec = redisClient.exec();
    log.info("EXEC {}", exec);

    return "done";
  }
}

```

### Expected behavior

Logging output of blocking and reactive code should be the same.

### Actual behavior

Logging output of blocking and reactive code is different.

### How to Reproduce?

See the following minimal project:

[redis.zip](https://github.com/quarkusio/quarkus/files/8168663/redis.zip)

### Output of `uname -a` or `ver`

Win

### Output of `java -version`

11

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

2.7.2.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

mvn 3.8.4

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by cescoffier**
The current Redis extension does not provide all the possibilities Redis offers, and the API is not easy to use.

This PR provides a new high-level and type-safe Redis API that covers the previous API's shortcomings.

The previous API is still available but deprecated.

**Breaking Changes**

Unfortunately, while refactoring the extension, I had to break a few things:

- Host Providers are now identified using `@Identifier` instead of `@Named` (following @mkouba's advice)
- It is not possible to create dynamic clients - this was done to support pub/sub, which is now supported by the API

**What's missing (and will come later)**

The PR is already big enough, and some parts will be provided later without breakage:

- the commands from the `stream` group are not yet provided (you can use custom commands for the time being)
- scan operations should provide a flattened stream (so instead of retrieving batches, retrieves the element individually) 
- metrics have been left off for now. It's possible to implement a micrometer binder (work in progress)
- the `LCS` command does not provide the `IDX` option

**Related issues**

- Fix https://github.com/quarkusio/quarkus/issues/24056
- Fix https://github.com/quarkusio/quarkus/issues/21108
-

### CODE REVIEW

**Comment by machi1990** [comment]
Thanks @cescoffier for this. Very nice improvements.

I'll have a proper look in the next couple of days.

**Comment by ebullient** [comment]
Is it too much to ask to bring the appropriate doc templates into your new docs? 
https://github.com/quarkusio/quarkus/pull/25354

docs/src/main/asciidoc/redis.adoc -- Is that meant to be a tutorial? (I believe so.. )
docs/src/main/asciidoc/redis-reference.adoc -- this feels like a combination of a how-to (installation), and a reference (how the extension works and config options, etc.) .. can we split them so we get the docs roughly better in the first pass? (I am happy to help in that regard.. )

File: docs/src/main/asciidoc/redis-reference.adoc [hunk 1]
```diff
@@ -3,99 +3,646 @@ This guide is maintained in the main Quarkus repository
...
+This API is fully reactive and asynchronous.
+It exposes a Mutiny API.
+It exposes the `io.quarkiverse.redis.datasource.api.ReactiveRedisDataSource` interface.
+- a _high-level_ imperative data source: A type-safe, high-level API to interact with Redis.
+It is the imperative variant of the reactive data source.
+It exposes the `io.quarkiverse.redis.datasource.api.RedisDataSource` interface.
+
+To help you select the suitable API for you, here are some recommendations:
+
+* If you are building an imperative (_classic_) Quarkus application integrating with Redis: use `io.quarkiverse.redis.datasource.api.RedisDataSource`.
```
Comment (machi1990) [inline docs/src/main/asciidoc/redis-reference.adoc:80]:
```suggestion
* If you are building an imperative (_classic_) Quarkus application integrating with Redis: use `io.quarkus.redis.datasource.api.RedisDataSource`.
```

The package name change to `io.quarkus.redis.datasource.api` has to be applied throughout 
```suggestion
* If you are building an imperative (_classic_) Quarkus application integrating with Redis: use `io.quarkiverse.redis.datasource.api.RedisDataSource`.
```

File: docs/src/main/asciidoc/redis-reference.adoc [hunk 2]
```diff
@@ -3,99 +3,646 @@ This guide is maintained in the main Quarkus repository
...
+quarkus.redis.client-type=SENTINEL
+
+# Optional
+quarkus.redis.master-name=my-sentinel # Default is my-master
+quarkus.redis.role=MASTER # Master is the default
+----
+
+=== Cluster Mode
+
+When using Redis in cluster mode, you need to pass multiple _host urls_, configure the client type to `CLUSTER` and configure the `replicas` mode:
```
Comment (machi1990) [inline docs/src/main/asciidoc/redis-reference.adoc:180]:
```suggestion
When using Redis in cluster mode, you need to pass multiple _host urls_, configure the client type to `cluster` and configure the `replicas` mode:
```

**Review by machi1990 — APPROVED** [review]
I left some minor doc suggestions, otherwise lgtm. 

Will it be possible to add a migration note as well as there are some breaking changes?

I love the direction that this is taking with typesafety, repository style api, thanks @cescoffier for the PR.

File: docs/src/main/asciidoc/redis-reference.adoc (hunk 1 above)
Comment (cescoffier) [reply docs/src/main/asciidoc/redis-reference.adoc]: Great catch! I initially planned to contribute the new client to quarkiverse.

**Pull request merged**